In [1]:
%run startup.py

# Problem 1

In [2]:
# Assign a friendly name to your network and snapshot
NETWORK_NAME = "example_network_nikita"
SNAPSHOT_NAME = "example_snapshot_nikita"

SNAPSHOT_PATH = "networks/example_Modified"

# Now create the network and initialize the snapshot
bf_set_network(NETWORK_NAME)
bf_init_snapshot(SNAPSHOT_PATH, name=SNAPSHOT_NAME, overwrite=True)

# retrieve the list of available batfish questions
load_questions()

## Question 1a. 

In [3]:
interface_properties = bfq.interfaceProperties(nodes="host.*").answer().frame()

In [4]:
interface_properties

Interface Access_VLAN Active Allowed_VLANs        All_Prefixes  \
0  host2:eth0  None        True                 ['2.128.1.101/24']   
1  host1:eth0  None        True                 ['2.128.0.101/24']   
2  host3:eth0  None        True                 ['2.128.2.101/24']   

  Auto_State_VLAN Bandwidth Blacklisted Channel_Group Channel_Group_Members  \
0  True            1e+09     False       None          []                     
1  True            1e+09     False       None          []                     
2  True            1e+09     False       None          []                     

  Declared_Names Description DHCP_Relay_Addresses Encapsulation_VLAN  \
0  ['eth0']       None        []                   None                
1  ['eth0']       None        []                   None                
2  ['eth0']       None        []                   None                

  HSRP_Groups HSRP_Version  \
0  []          None          
1  []          None          
2  []          None          

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   Incoming_Filter_Name  \
0  IpAccessList:filter::INPUT\nIpAccessListLine{action=PERMIT, matchCondition=MatchHeaderSpace{headerSpace=HeaderSpace{dstPorts=[[22,22]], ipProtocols=[TCP]}}, name=-p tcp --dport 22 -j ACCEPT}\nIpAccessListLine{action=DENY, matchCondition=TrueExpr{}, name=default}                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 
1  IpAccessList:filter::INPUT\nIpAccessListLine{action=PERMIT, matchCondition=MatchHeaderSpace{headerSpace=HeaderSpace{dstPorts=[[53,53]], ipProtocols=[UDP]}}, name=-p udp --dport 53 -j ACCEPT}\nIpAccessListLine{action=PERMIT, matchCondition=MatchHeaderSpace{headerSpace=HeaderSpace{dstPorts=[[22,22]], ipProtocols=[TCP]}}, name=-p tcp --dport 22 -j ACCEPT}\nIpAccessListLine{action=DENY, matchCondition=TrueExpr{}, name=default}                                                                                                                                                                                                                                                                                                                                             
2  IpAccessList:filter::INPUT\nIpAccessListLine{action=PERMIT, matchCondition=MatchHeaderSpace{headerSpace=HeaderSpace{dstPorts=[[21,21]], ipProtocols=[UDP]}}, name=-p udp --dport 21 -j ACCEPT}\nIpAccessListLine{action=PERMIT, matchCondition=MatchHeaderSpace{headerSpace=HeaderSpace{dstPorts=[[22,22]], ipProtocols=[TCP]}}, name=-p tcp --dport 22 -j ACCEPT}\nIpAccessListLine{action=PERMIT, matchCondition=MatchHeaderSpace{headerSpace=HeaderSpace{dstPorts=[[161,161]], ipProtocols=[UDP]}}, name=-p udp --dport 161 -j ACCEPT}\nIpAccessListLine{action=DENY, matchCondition=MatchHeaderSpace{headerSpace=HeaderSpace{dstPorts=[[161,161]], ipProtocols=[TCP]}}, name=-p tcp --dport 161 -j DROP}\nIpAccessListLine{a

Host 3 allows UDP traffic to port 161, whereas the other hosts do not allow that traffic.The SNMP agent receives requests on UDP port 161. So, host3 is the one that can handle SNMP requests.

## Question 1b. 

In [9]:
path = PathConstraints(startLocation=".*")
headers = HeaderConstraints(dstIps="ofLocation(host3)")
reach = bfq.reachability(pathConstraints=path, headers=headers, actions="success").answer().frame()
show(reach)

Flow  \
0   start=as1border1 [1.0.1.1:49152->2.128.2.101:22 TCP]   
1   start=as1border2 [1.0.2.1:49152->2.128.2.101:22 TCP]   
2   start=as1core1 [1.0.1.2:49152->2.128.2.101:22 TCP]     
3   start=as2border1 [2.1.1.1:49152->2.128.2.101:22 TCP]   
4   start=as2border2 [2.1.1.2:49152->2.128.2.101:22 TCP]   
5   start=as2core1 [2.1.2.1:49152->2.128.2.101:22 TCP]     
6   start=as2core2 [2.1.2.2:49152->2.128.2.101:22 TCP]     
7   start=as2dept1 [2.1.10.2:49152->2.128.2.101:22 TCP]    
8   start=as2dist1 [2.1.3.1:49152->2.128.2.101:22 TCP]     
9   start=as2dist2 [2.1.3.2:49152->2.128.2.101:22 TCP]     
10  start=as3border1 [3.0.1.1:49152->2.128.2.101:22 TCP]   
11  start=as3border2 [3.0.2.1:49152->2.128.2.101:22 TCP]   
12  start=as3core1 [3.0.1.2:49152->2.128.2.101:22 TCP]     
13  start=host1 [2.128.0.101:49152->2.128.2.101:22 TCP]    
14  start=host3 [2.128.2.101->2.128.2.101 ICMP]            

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   

In [11]:
path = PathConstraints(startLocation="host2")
headers = HeaderConstraints(dstIps="ofLocation(host3)")
reach = bfq.reachability(pathConstraints=path, headers=headers, actions="success").answer().frame()
show(reach)

Empty DataFrame
Columns: [Flow, Traces, TraceCount]
Index: []

In [10]:
path = PathConstraints(startLocation=".*")
headers = HeaderConstraints(dstIps="ofLocation(host3)")
reach = bfq.reachability(pathConstraints=path, headers=headers, actions="failure").answer().frame()
show(reach)

Flow  \
0   start=as1border1 [1.0.1.1->2.128.2.101 ICMP]   
1   start=as1border2 [1.0.2.1->2.128.2.101 ICMP]   
2   start=as1core1 [1.0.1.2->2.128.2.101 ICMP]     
3   start=as2border1 [2.1.1.1->2.128.2.101 ICMP]   
4   start=as2border2 [2.1.1.2->2.128.2.101 ICMP]   
5   start=as2core1 [2.1.2.1->2.128.2.101 ICMP]     
6   start=as2core2 [2.1.2.2->2.128.2.101 ICMP]     
7   start=as2dept1 [2.1.10.2->2.128.2.101 ICMP]    
8   start=as2dist1 [2.1.3.1->2.128.2.101 ICMP]     
9   start=as2dist2 [2.1.3.2->2.128.2.101 ICMP]     
10  start=as3border1 [3.0.1.1->2.128.2.101 ICMP]   
11  start=as3border2 [3.0.2.1->2.128.2.101 ICMP]   
12  start=as3core1 [3.0.1.2->2.128.2.101 ICMP]     
13  start=host1 [2.128.0.101->2.128.2.101 ICMP]    
14  start=host2 [2.128.1.101->2.128.2.101 ICMP]    

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           

Feasible paths exist from A1B1, A1B2, A1C1, A2B1, A2B2, A2C1, A2C2, A2De1, A2Di1, A2Di2, A3B1, A3B2, A3C1, Host1, Host3 to the SNMP host. Host2 is the router that does not have any viable path to get to host3.

## Problem 1 c.  

All the routes that pass through the router A2C2 have a loop on them. The traffic is passed back from A2C2 to either A2B1 or A2B2, and comes back to A2C2 and repeats over and over.

One way to ensure that there is no loop, is to make sure that the traffic heading from A2B2/A2B1 to host3 via A2C2, goes to A2C1 instead. 

## Problem 1 d. 

I think that intuitively, this fix should cause optimal routes. Considering that all traffic to host3 must route through A2C1 post this fix, the new route should be an optimal path.

## Problem 1 e.

1)Don't let control packets inform AS2 of the presence of AS1 and AS3 
2)Alter the ACL such that SNMP services cannot reach AS1 and AS3
3)First one is preferred because there are more tools to help control plane than for the data plane issues